modules/core/shared/src/main/scala/net/protocol/Close.scala

```diff
@@ -20,27 +20,40 @@ trait Close[F[_]] {
 object Close {
 
   def apply[F[_]: FlatMap: Exchange: MessageSocket: Tracer]: Close[F] =
-    new Close[F] {
-
-      override def apply(portalId: Protocol.PortalId): F[Unit] =
-        exchange("close-portal") { (span: Span[F]) =>
-          span.addAttribute(Attribute("portal", portalId.value)) *>
-          close(CloseMessage.portal(portalId.value))
-        }
-
-      override def apply(statementId: Protocol.StatementId): F[Unit] =
-        exchange("close-statement") { (span: Span[F]) =>
-          span.addAttribute(Attribute("statement", statementId.value)) *>
-          close(CloseMessage.statement(statementId.value))
-        }
-
-      def close(message: CloseMessage): F[Unit] =
-        for {
-          _ <- send(message)
-          _ <- send(Flush)
-          _ <- expect { case CloseComplete => }
-        } yield ()
+    new AbstractClose[F] {
+      protected def weakExchange[A](label: String)(f: Span[F] => F[A]) =
+        exchange(label)(f)
+    }
 
+  /** Like [[apply]] but doesn't acquire a mutex, allowing usage from within an existing exchange. */
+  private[skunk] def midExchange[F[_]: FlatMap: MessageSocket: Tracer]: Close[F] =
```

Comment:
Total hack to re-use the logic in closing a statement from within the `ParseDescribe` exchange without acquiring the session mutex and hence deadlocking.